### DISCUSSION TITLE
Running Powershell Script

### DISCUSSION THREAD

**Discussion by AndrewSmithDev** [AutoDarkMode/Windows-Auto-Night-Mode#456] [Q&A] [answered]
I'm trying to run a Powershell script using the [custom script](https://github.com/AutoDarkMode/Windows-Auto-Night-Mode/wiki/How-to-add-custom-scripts) feature but I can't figure out why it's not working.

I can run the following commands from cmd and the script works

```
C:\>powershell.exe -file C:\toggle-win-terminal-theme.ps1 "One Half Light" "One Half Dark"
C:\>powershell.exe -file C:\toggle-win-terminal-theme.ps1 "One Half Dark" "One Half Light"
```

However, when I try to run a custom script in Auto Night Mode it doesn't work. Here is my `scripts.yaml` file

```Enabled: false
Component:
  TimeoutMillis: 10000
  Scripts:
    - Name: WindowsTerminal
      Command: powershell.exe
      ArgsLight: [-file, C:\toggle-win-terminal-theme.ps1, "One Half Dark", "One Half Light"]
      ArgsDark: [-file, C:\toggle-win-terminal-theme.ps1, "One Half Light", "One Half Dark"]
```

**Comment by Spiritreader** [comment, marked as answer]
you want sth like this

````
  - Name: pwsh example
    Command: powershell 
    ArgsLight: [C:\Users\User\AppData\Roaming\AutoDarkMode\test.ps1, -message 'Hello World!' -emotion 'happy']
    ArgsDark: [C:\Users\User\AppData\Roaming\AutoDarkMode\test.ps1, -message 'Hello World!' -emotion 'happy']
    AllowedSources: [Any]
````
You're basically calling powershell and then want to pass the script path as an argument, and your actual arguments as another argument

Powershell is quite finicky in terms of how it wants to be called.
ADM wraps all arguments, so with your script configuration example, you're calling
`"powershell.exe" "-file" "C:\toggle-win-terminal-theme.ps1", """One Half Light"", """One Half Light"""`
which will not work.

**Reply by AndrewSmithDev** [reply]
Since Powershell is wonky I decided to just rewrite the script in node. I was using someone else's script that I found on GitHub but I'm not familiar with Powershell at all so I decided to just write my own version. It was much more straight forward getting the script to work with node.

```yaml
Component:
  TimeoutMillis: 10000
  Scripts:
    - Name: WindowsTerminal
      Command: node
      ArgsLight:
        - C:\toggle-win-term-color.js
        - -light
      ArgsDark:
        - C:\toggle-win-term-color.js
        - -dark
      AllowedSources: [Any]
```

---

This is a bit off topic but I wrote a script that will toggle the color theme of Windows Terminal. Here is the script for anyone interested

```javascript
const { readFileSync, writeFileSync } = require("fs");
const { join } = require("path");

const colorConfigs = [
  { dark: "Campbell", light: "Campbell Powershell" },
  { dark: "One Half Dark", light: "One Half Light" },
  { dark: "Solarized Dark", light: "Solarized Light" },
  { dark: "Tango Dark", light: "Tango Light" },
];

const modeArg = process.argv[2];
if (modeArg !== "-dark" && modeArg !== "-light") {
  console.error(`Invalid Argument. Argument must be '-dark' or '-light'. Received '${modeArg}'`);
  process.exit(-1);
}

const isDarkMode = modeArg === "-dark" ? true : false;

const pathToWinTermConfig = join(
  process.env.LocalAppData,
  "/Packages/Microsoft.WindowsTerminal_8wekyb3d8bbwe/LocalState/settings.json"
);

const configContents = readFileSync(pathToWinTermConfig).toString("utf-8");
const config = JSON.parse(configContents);

config.profiles.list.forEach((profile) => {
  const config = colorConfigs.find((c) => profile.colorScheme === c[isDarkMode ? "light" : "dark"]);
  if (config) profile.colorScheme = config[isDarkMode ? "dark" : "light"];
});

writeFileSync(pathToWinTermConfig, JSON.stringify(config, null, 2));
```

**Reply by Spiritreader** [reply]
Thanks for sharing!

I've added the example I mentioned above to the wiki nonetheless in case someone else wants to get powershell running!
